### DISCUSSION TITLE
Inverse operator for ??

### DISCUSSION THREAD

**Discussion by andrey-messi** [dotnet/csharplang#7855] [General]
I think this is a duplicate request, but..
If we have this code^
W q = w != null ? w : new W();
then we can shorten it in this code:
W q = w ?? new W();
But I often need to make logic that inversed, I need this:
Y q = w == null ? null : new Y();
It is very long expression and need to be shortened. Every time I hope in new version C# we will have something like 
Y q = w !?  new Y();
but when I try it in my current net8 project it give compile error.

Whar the problem to realize this function 
Y q = w !?  new Y();
?

**Comment by Richiban** [comment]
Sorry, but I'm really struggling to see where this would ever be useful; you want reassign a variable only if it's *not* null?

Can you give a more concrete or real-world example of where this might be useful?

**Reply by TheUnlocked** [reply]
Their request is similar to Option.map in languages that have that. A common use of that is to get members of something that could be null (which is what the `?.` operator already does), but sometimes it can be useful to run a function on the value, preserving nulls.

With that said, I haven't had a particular need for this in C#. I do often find myself wanting a similar operator in JS/TS, but extension methods paired with `?.` cover a lot of use cases in C#, and C#'s nulls aren't nearly as flexible as in JS anyways.

**Reply by suugbut** [reply]
Please kindly see my proposal. https://github.com/dotnet/csharplang/discussions/7855#discussioncomment-8272641

**Reply by dersia** [reply]
I on the other hand have a lot of use for this feature in recent development.

**Comment by jnm2** [comment]
The specific syntax `a !? b` would be complicated to do because it's the same as `a! ? b` which might already appear in C# as part of a conditional expression. Not impossible, but not appealing. ([SharpLab](https://sharplab.io/#v2:EYLgtghglgdgNAFxANwKYCcoDMCeAfAAQCYBGAWAChLkJ0ACCOgXjplQHc6AxAex4AoAlAG5qtOsGasO3PkNEUa9AMZS2nXgJGVKAfSmMAhAH4JdEHWULKxWT0oBvSnRd0CAZjckAbBL4AbOh4ABwwIBB56BHQAV1R+TTosPkFmAD46BAALdB5OdToAOR4EAEkwYP9UMFQYBFQAEwBRAA9lVGCEKB4YeWdXDy9fYACg0PRwyKSIfwBneMTknlSmDOzc/JlisoqqmrrG1vbO7t6RVzpKAF9KIA===))

**Reply by andrey-messi** [reply]
Maybe this syntax?

a !?? b

**Reply by HaloFour** [reply]
`a! ?? expr` is also already valid syntax.

**Reply by andrey-messi** [reply]
What does "!" in "a! ?? expr" mean? As I know,  in that case "!" after "a" just tell a static analizer that "a" is not null to suppress showing warnings. But guaranteed not null "a" has no sense when you need
Y q = a !?? new Y(a);
So I think it is possible to use "!??"

**Reply by HaloFour** [reply]
It is the null forgiveness operator and does behave as you have described.  But since it is already legal code today that cannot be changed.

**Reply by HaloFour** [reply]
Syntax aside, I think the value of having any language feature here has to be demonstrated.  I don't think it's a remotely common pattern and the existing code via ternary conditional operator is plenty terse.

A `map` like operator could be interesting, but it's useless unless it has access to the unwrapped value.

**Reply by andrey-messi** [reply]
"existing code via ternary conditional operator is plenty terse." - so in the next version of C# we should remove ?? ...

" I think the value of having any language feature here has to be demonstrated. "
for example. i have logic with car and garage . If i have car i need garage, if i dont have car i dont need garage.

What can I do?
1)
Garage q = car  == null ? null : new Garage();
2)
Garage q;
if(car  != null)
 q = new Garage();

What I want:
Garage q = car  !?? new Garage();

Because in real coding it is not rare than "W q = w != null ? w : new W();" case that has ?? operator.

**Reply by HaloFour** [reply]
Wanting a default fallback for a `null` value is a very common occurrence and justified the language feature.  I don't believe this request is remotely as common, thus it does not justify a language feature specifically to support.  I don't see why either of the two alternatives you've presented would be considered onerously verbose.

**Comment by suugbut** [comment]
I propose the following syntax:

- `x ?< y` equals to  `x is null ? y : x` equals to the existing `x ?? y` 
- `x ?> y` equals to  `x is null ? x : y` 

### How to memorize? 
If the left expression `x` is `null` for

- `x ?< y` then `y` will be applied as Pacman's mouth is facing to `y`. 
- `x ?> y` then `x` will be applied as Pacman's mouth is facing to `x`. 


Non-trivial examples for `x ?> y` cases:

## Case 1

Buy a car only if we have a garage.

```csharp
Garage? garage = null;

Car? car = garage is null ? null: new Car(); // normal code
Car? car = garage ?> new Car(); // proposed alternate

class Car;
class Garage;
```



## Case 2

`isNut` is
- `true` if `fruit` is `Nut`
- `null` if `fruit` is `null`
- `false` if `fruit` is neither `Nut` nor `null`

```csharp
Fruits? fruit = null;

bool? isNut = fruit is null ? null : fruit == Fruits.Nut; // normal code
bool? isNut = fruit?.Equals(Fruits.Nut); // normal code
bool? isNut = fruit ?> fruit == Fruits.Nut; // proposed alternate

public enum Fruits
{
    Nut,
    Kiwi,
    Melon
}
```

Note: For my previous proposed syntax, see https://github.com/dotnet/csharplang/discussions/7881#discussion-6140773

## Contemplation
For the sake of fairness, let's consider the following:

- Why do we have `x ?? y` when we can just write `x is null ? y : x` that is more expressive, readable, idiomatic?
- Why do we have `x?.Something` when we can just write `x is null ? null : x.Something` that is more expressive, readable, idiomatic?
- To be honest, all of these expressions `x ?? y`,  `x ?< y`,  `x ?> y` have the same degree of "obscurity" for new comers until they learn and become familiar with.

**Reply by tacosontitan** [reply]
```cs
Car? car = garage ?> new Car();
```
This hides the fact that `null` is being assigned if `garage` is `null` IMHO.

**Reply by suugbut** [reply]
Yes. It does make sense. If you don't have a garage, you should not buy a car.

**Reply by tacosontitan** [reply]
I think you misunderstood my point; I'm saying that the snippet doesn't explicitly demonstrate that `null` is a possibility for assignment to `car`.

**Reply by HaloFour** [reply]
I'd much rather the code be written to explicitly spell that out.

**Reply by suugbut** [reply]
To be honest, both `x ?? y` or `x ?< y` have the same degree of "obscurity" for new comers until they learn and become familiar with.

**Reply by suugbut** [reply]
But `car` is of type `Car?`. It is clear that `null` is possible for `car`.

**Reply by tacosontitan** [reply]
> To be honest, both `x ?? y` or `x ?< y` have the same degree of "obscurity" for new comers until they learn and become familiar with.

That's not equivalent at all. In `a = b ?? c` it is *explicitly* clear that ***only*** `b` *or* `c` will be assigned to `a`, regardless of what `??` actually means. In `a = b ?< c`, it is explicit that `b` or `c` can be assigned but `null` can also be assigned, making it implicit behavior.

**Reply by suugbut** [reply]
It is not implicit behavior.
- The type of `a` in either `a = b ?< c` or `a = b ?? c` is not nullable.
   ```csharp
   // normal scenario
   int c = 1;
   int? b = null;
   int a = b ?? c;
   int a = b ?< c; // equivalent
   ```

- The type of `a` in `a = b ?> c` is nullable.
   ```csharp
   // normal scenario
   int c = 1;
   int? b = null;
   int? a = b ?> c;
   ```

**Reply by viceroypenguin** [reply]
I'm gonna be honest, `Car? car = garage ?> new Car();` is problematic to me for an entirely different reason. 

With `a = b ?? c`, you always have the explicit agreement that `a = b` and `a = c` are valid statements. You know that you could assign either `b` or `c` to `a`. The operator is shortening _which one_.

With `Car? car = garage ?> new Car();` you break that assumption. You can, at no time, ever do `Car? car = garage` (if `Car` and `Garage` are independent types with no assignment relation). Even if `garage == null`, you still can't do `car = garage`, because they are independent types. Adding `?>` capability of doing so would break all sorts of readability expectations.

**Reply by suugbut** [reply]
New comers who read `int? id = person?.Id;` also might get confused how `person` of type `Person?` can be assigned to `id` of type `int?` when `person = null`, there is a mismatch type here. It really breaks all sorts of readability expectations. It does not obey **The Principle of Least Surprise (POLS)**. 😊

```csharp
Person? person = null;

int? id = person?.Id;

class Person
{
    public int Id { get; set;}
}
```

So actually behind the scene, the compiler can do something as follows. 

```csharp
Garage? garage = null;
Car? car = garage ?> new Car();
Car? car = garage is null ? (Car?) garage: new Car();
Car? car = garage is null ?  null : new Car();
```

Getting familiar with new things can only be achieved by learning.

**Comment by tacosontitan** [comment]
I'm against this idea because it seems like a request to support writing shorter code for the sake of writing shorter code, especially given your statement:

> But I often need to make logic that inversed, I need this:
> ```cs
> Y q = w == null ? null : new Y();
> ```
> It is *very long expression and need to be shortened*.

<sub>*Emphasis mine.*</sub>

The use case `q = w is null ? null : new Y()` just screams "refactor needed", and without context (to me at least), doesn't really make any sense. I second @Richiban's question:

> Can you give a more concrete or real-world example of where this might be useful?

I do like the fact that you made a comment to another poster using the car and garage example, but it still left me asking the question "why are you creating a variable you don't need?". Everything I came up with around this ended in me refactoring it to something more legible than creating an unnecessary variable. For example:

```cs
Car? car = await service.GetCar(id: 12345, cancellationToken);
if (car is null)
    return;

Garage garage = new(car);
...
```

**Reply by suugbut** [reply]
A more reasonable code snippet.
```csharp
public async Task<Whatever> TryDeliveryingCarAsync(int tenantId)
{
      Garage? garage = await service.GetGarageByTentantId(id: tenantId, cancellationToken);
      Car? carPrize = garage is null ? null : new Car();
       
      // or using the proposed operator
      Car? carPrize = garage ?> new Car();
       
      return await DoSomethingElseAsync(carPrize, tenantId);
 }
```

**Reply by HaloFour** [reply]
I don't find the version without this proposed operator to be onerous.  It's clear and idiomatic.

**Reply by suugbut** [reply]
Yes. I agree. But sorry, 

`x ?? y` should not be  implemented too because  `x is null ? y : x` is not onerous and it is clear and idiomatic.

**Reply by HaloFour** [reply]
You're welcome to your opinion, but the team thought that wanting to replace a `null` with another value was common and valuable enough to warrant a specific language feature.  I would suggest your use case is several orders of magnitude less common.

**Reply by tacosontitan** [reply]
@suugbut, If I've offended you in any way by disagreeing with your proposed syntax then I apologize, that's certainly not my intent. I simply wanted to point out that it wasn't explicit enough that `null` was a potential outcome, in my opinion.

However, since you feel the need to imply that my snippet is *unreasonable* and provided a snippet that is "more reasonable", let me ask why you're creating an instance of `Car?` when you're not using it at all locally; that's something that is useless in your snippet and should be the responsibility of `DoSomethingElseAsync` instead.

I'm curious though, this is how I would write the code you're proposing as "more reasonable":

```cs
public async Task<DeliveryResult> DeliverCarToTenant(Tenant tenant, CancellationToken cancellationToken)
{
    Garage? garage = await garageService.GetGarageByTentant(tenant, cancellationToken);
    if (garage is null)
        return DeliveryResult.TenantNotRegisteredWithGarage;

    Car? car = await carService.GetTenantCarFromGarage(tenant, garage, cancellationToken);
    if (car is null)
        return DeliveryResult.CarNotFound;

    return await deliveryService.DeliverCarToTenant(car, tenant cancellationToken)
 }
 ```

How would you refactor this using your proposal to make it more legible than it already is?

**Reply by suugbut** [reply]
I think there are many other non-trivial examples for ` x ?> y` use cases.

```csharp
public async Task<DeliveryResult> DeliverCarToTenant(Tenant tenant, CancellationToken cancellationToken)
{
    Garage? garage = await garageService.GetGarageByTentant(tenant, cancellationToken);

    Car? car = await carService.GetTenantCarFromGarage(tenant, garage, cancellationToken);
    if (car is null) return DeliveryResult.CarNotFound;

    return await deliveryService.DeliverCarToTenant(car, tenant cancellationToken)
 }

class CarService(AppDbContext context, ...)
{
    public async Task<Car?> GetTenantCarFromGarage(Tenant tenant, Garage? garage, CancellationToken cancellationToken)
    {
        return garage is null ? null : await WhateverGetCar();
        return garage ?> await WhateverGetCar(); 
    }
}
```

**Reply by tacosontitan** [reply]
I think we're going around in circles, so my final correspondence will be:

Why are you not just guarding against the incoming `garbage` being `null` and returning early?

```cs
class CarService(AppDbContext context, ...)
{
    public async Task<Car?> GetTenantCarFromGarage(tenant, garage, cancellationToken)
    {
        if (garage is null || tenant is null)
            return null;

        // From where a car is retrieved is not important.
        // For simplicity, let's create it manually with new Car().
        await SomethingElseAsync();
        return new Car(); // use either this line
        return new Car(); // or this line
    }
}
```

**Reply by dersia** [reply]
Honestly I can't say that I like any of the syntax proposed here, but I have been seeing myself typing 
`var x = y is null ? null : y` very often recently and I wished for a quicker syntax like we have with `??`.

So I would support having a feature that does the invert of what `??`, but I don't like any of the syntax I have seen so far in this discussion.

**Reply by suugbut** [reply]
@dersia : Or do you have any idea for the operator?

how about the following?

- ` x ?| y` for `x is null ? y : x` equals to `x ?? y`
- ` x |? y` for `x is null ? x : y `

**Reply by dersia** [reply]
honestly I thought about it a bit more, I am still not sure, the only thing I could do was looking at it reasonably and this is what I came up with:

`?` means null-check
`!` means that's OK, I know what I am doing
so based on this idea I would suggest `?!`

but again, I am not sure if this is a good compromise or even a good operator at all, I just wanted to say, that I often catch myself wishing for an operator that does this, but I can also live without it. Just if in this discussion we are looking for if there are developers that could use this feature, consider me as someone who wants this 😊

**Reply by suugbut** [reply]
As already mentioned in https://github.com/dotnet/csharplang/discussions/7855#discussioncomment-8171037, your proposed operator `x ?! y` can also have another meaning `x ? !y`.

**Reply by andrey-messi** [reply]
"but it still left me asking the question "why are you creating a variable you don't need?". Everything I came up with around this ended in me refactoring it to something more legible than creating an unnecessary variable."

For example I create a model on backend which i need to return on frontend.

```
var car = GetSomeCar();
return new CarGarageBlahBlahBlahData()
{
  Car = car , 
  Garage = car !! new Garage(),
  Prop3 = 333, 
  ...
  Prop20 = 202020
}
```

I dont want to write 

```
var car = GetSomeCar();
if(car is null)
  return new CarGarageBlahBlahBlahData()
  {
    Car = car , 
    Garage = null,
    Prop3 = 333, 
    ...
    Prop20 = 202020
  }
else
  return new CarGarageBlahBlahBlahData()
  {
    Car = car , 
    Garage = new Garage(),
    Prop3 = 333, 
    ...
    Prop20 = 202020
  }
```

or if i write this code then again i create a "not always nessesery variable garage" 

```
var car = GetSomeCar();
Garage garage = null;
if(car is not null)
  garage = new Garage();

return new CarGarageBlahBlahBlahData()
{
  Car = car , 
  Garage = garage ,
  Prop3 = 333, 
  ...
  Prop20 = 202020
}
```

**Reply by HaloFour** [reply]
Ternary operator seems perfectly fine for a case like this.

```cs
var car = GetSomeCar();
return new CarGarageBlahBlahBlahData()
{
  Car = car , 
  Garage = car != null ? new Garage() : null,
  Prop3 = 333, 
  ...
  Prop20 = 202020
}
```

**Reply by andrey-messi** [reply]
"Ternary operator seems perfectly fine for a case like this."

but

"It is very long expression and need to be shortened. Every time I hope in new version C# we will have something like
Y q = w !? new Y();
but when I try it in my current net8 project it give compile error."

**Reply by HaloFour** [reply]
> It is very long expression and need to be shortened.

It's not a long expression, and it doesn't need to be shortened.  It's not a goal to make every possible expression in the language as terse as possible, nor is it a goal to invent new operators for every possible combination of logic that one could want to describe, regardless of whether or not they want to embed it in an initializer.

**Reply by tacosontitan** [reply]
@andrey-messi why not just set `Garage` conditionally after creating the new instance?
```cs
var car = GetSomeCar();
var blahBlahData = new CarGarageBlahBlahBlahData
{
  Car = car , 
  Prop3 = 333, 
  ...
  Prop20 = 202020
};

if (car is not null)
    blahBlahData.Garage = new Garage();

return blahBlahGarage;
```

**Reply by andrey-messi** [reply]
Because  "It is very long expression and need to be shortened"

Same question: why will you use 
Car car = car ?? new Car()
Instead of
Car car = car != null ? car : new Car()

?

**Reply by HaloFour** [reply]
Because wanting to be able to have a fallback for a `null` value is an infinitely more common operation to want to do, and the team found that it was worth a special operator.  Just because one operation gets a special operator does not mean that all operations are fair game for special operators.  This has been explained several times already.

**Reply by tacosontitan** [reply]
@andrey-messi, I feel as though I've really beaten my point into the ground at this point, so this will be my last response to this discussion as a whole because I have no interest in this feature.

Being terse, for the sake of being terse is never a good idea. This is a niche use case that really doesn't have a method (in my opinion) to explicitly represent intent and outcome in a more concise manner than a ternary, and it's better in those scenarios to simply be more verbose for the sake of legibility. A few other points that I've been really clear about avoiding are:

- Don't create variables you don't need.
- Don't assign to properties if you don't need to.
  - Why would you explicitly assign `null` if that's the `default` for your type?

My stance remains unchanged in that the many things I've mentioned are much more legible than even using a ternary to solve this problem.

- Apply inversion with a guard clause and early return.
- Assign to the property conditionally before returning.

I wish you the best of luck!

**Reply by andrey-messi** [reply]
I dont tell that "inverse ?? Operator" is a thing that i strongly need. What i want is just a syntax sugar for that logic.
And by my statistics i use case
A a = b == null ? Null : new A()
Not rare but maybe even often than
A a = a == null ? New A() : a
For that reason why should i think that my cases are "This is a niche use case"?

So if to want null is "a niche use case" then why we have this syntax sugar

Garage g = car?.ReturnGarageForThisCar();

We can always do that:
Garage g = car==null ? null : car.ReturnGarageForThisCar();

**Reply by suugbut** [reply]
@andrey-messi :  Out of topic. To make your code well formatted, use the following template.
\`\`\`csharp
// your code goes here
\`\`\`
where \` is a backtick (a key to the left of 1 on our QWERTY keyboard).

**Reply by HaloFour** [reply]
> then why we have this syntax sugar

The answer is going to always be that it was found to be a common enough scenario to warrant a language feature.  The desire to navigate deep call chains without nested conditional statements is common enough that the null propagation operator exists in quite a few programming languages.

This case is much less common, and the suggested behavior of such a proposed operator to be very unintuitive.

**Reply by Daynvheur** [reply]
There is a difference in `??` vs `!?` (or whatever other inverted character I can't have without a numeric pad):
In favor of `??`:
```cs
int a = GetIntFromDatabase() != null ? GetIntFromDatabase() : -1; // This needs a method call refactor

int? b = GetIntFromDatabase(); // This is a method call refactor
int aBis = b != null ? b : -1; // This is equivalent to int aBis = b ?? -1 // This is equivalent to a = GetIntFromDatabase() ?? -1;
```

In defavor of `!?`:
```cs
int? a = GetIntFromDatabase() != null ? -1 : null; // This doesn't need a method call refactor

int? b = GetIntFromDatabase() !? -1; // This isn't shortening that much, and hides the defaultening intent
```

**Comment by suugbut** [comment]


**Reply by tacosontitan** [reply]
I'm not sure why you felt the need to edit your post to make it seem like people are voting on your ideas, but I'll go ahead and reassure others with this comment that my 👎 on this is not a vote for any of these proposals.

I disagree with all of them because none of them are explicit that `null` is a possibility for assignment to the target in `a = x OP y`.

I think we've run this particular topic into the ground though, so I will no longer be responding to your comments on it in an effort to avoid further frustration on my end, and yours.

Best of luck!

**Reply by viceroypenguin** [reply]
@suugbut I'm in the same boat. You are adding more items to make it appear as if the existing votes are in favor of the certain ideas, when they are not. Please stop attempting to manipulate results like this.

**Reply by CyrusNajmabadi** [reply]
My downvote is an actual downvote.  not an actual vote for the second line.  I don't like any of thse.

**Reply by suugbut** [reply]
The table was designed from the beginning for voting and I am updating this table gradually.

**Reply by dersia** [reply]
my 🫤 was not for the option that I just mentioned because there was a ask, I don't proposed that solution nor was my emote ment to be used for that. I just chose an emote that wasn't an option at that time to express that I don't vote for any of the options. so if you'd be so nice to remove me from the list or remove that option completely, I'd appreciate it

**Reply by suugbut** [reply]
I have made a new one to avoid misunderstanding. https://github.com/dotnet/csharplang/discussions/7855#discussioncomment-8340405 I hope the given information at the top is clear enough. Thank you.

**Comment by dersia** [comment]
long story short, it would be nice to have an operator for this, but I don't feel strongly about it. if maintainers think we don't need it, it's fine with me, there are more important features that I am looking forward to like extions, unions, expression bodies and HKT (I know it's unlikely that we get HKT, but I am still hoping)

**Comment by suugbut** [comment]
This "answer" is made only for ones supporting the proposal. 
If you intent is for rejection, please react to the existing posts or make new ones.

| Vote | `x is null ? y : x` | `x is null ? x : y` | by |
| ---------|----------|----------|-------------|
| 👍 | `x ?\| y` | `x \|? y` | suugbut |
| 👎 | `x ?? y` | `x ¿¿ y` | suugbut |
| ❤️ | `x ?< y` or `x <? y` | `x ?> y` or `x >? y`| suugbut |
| 🎉 | `x ?? y`  |  `x !? y` | andrey-messi **(*)** |
| 🚀 | `x ?? y` | `x ⸮⸮ y` | huoyaoyuan **(*)** |
| 👀 |  |  |
| 🚀 |  |  |
| 😕 |  |  | |
| 😂 |  |  | |

**(*)** : Feel free to ask for attribution removal.

**Reply by HaloFour** [reply]
None of the above, this operation isn't remotely common enough to justify a special operator.

**Reply by Daynvheur** [reply]
Maybe reconsider the usage of inverted characters, as I don't (and don't plan to) have access to inverted characters,
I comment here to maybe narrow your proposal list.

**Reply by suugbut** [reply]
Yes. Both the inverted and mirrored ones are not in 7-bit ASCII characters. I keep putting them in the table only for the sake of documentation. In the future other people who come with the same idea should not propose them again.
